File: strategy-pattern/text-editor/text-editor.py
from abc import ABC, abstractmethod


# Strategy Interface definition
class TextFormatter(ABC):
    @abstractmethod
    def format(self, text: str) -> str:
        pass


# --- Concrete Strategies ---
class PlainTextFormatter(TextFormatter):
    def format(self, text: str) -> str:
        return text


class BoldTextFormatter(TextFormatter):
    def format(self, text: str) -> str:
        return f"**{text}**"


class ItalicTextFormatter(TextFormatter):
    def format(self, text: str) -> str:
        return f"*{text}*"


# Context Class
class TextEditor:
    def __init__(self, formatter: TextFormatter):
        self.formatter = formatter

    def set_formatter(self, formatter: TextFormatter):
        self.formatter = formatter

    def display(self, text: str):
        print(self.formatter.format(text))


# Usage
inputText = "Hello, World!"

editor = TextEditor(PlainTextFormatter())
editor.display(inputText)

editor.set_formatter(BoldTextFormatter())
editor.display(inputText)

editor.set_formatter(ItalicTextFormatter())
editor.display(inputText)

File: strategy-pattern/text-editor/text-editor.ts
interface TextFormatter {
  format(text: string): string;
}

// <----------- Concrete strategies ------------>
class PlainTextFormatter implements TextFormatter {
  format(text: string): string {
    return text;
  }
}

class BoldTextFormatter implements TextFormatter {
  format(text: string): string {
    return `**${text}**`;
  }
}

class ItalicTextFormatter implements TextFormatter {
  format(text: string): string {
    return `*${text}*`;
  }
}

// Context class
class SimpleTextEditor {
  private formatter: TextFormatter;

  constructor(formatter: TextFormatter) {
    this.formatter = formatter;
  }

  setFormatter(formatter: TextFormatter) {
    this.formatter = formatter;
  }

  display(text: string) {
    console.log(this.formatter.format(text));
  }
}

// Usage
const inputText = "Hello from Typescript!!";

const simpleEditor = new SimpleTextEditor(new PlainTextFormatter());
simpleEditor.display(inputText);

simpleEditor.setFormatter(new BoldTextFormatter());
simpleEditor.display(inputText);

simpleEditor.setFormatter(new ItalicTextFormatter());
simpleEditor.display(inputText);
